### ISSUE TITLE
OpenRouter chat doesn't work with some providers/models

### ISSUE THREAD

**Issue by priyashpatil** [continuedev/continue#3378]
### Before submitting your bug report

- [X] I believe this is a bug. I'll try to join the [Continue Discord](https://discord.gg/NWtdYexhMs) for questions
- [X] I'm not able to find an [open issue](https://github.com/continuedev/continue/issues?q=is%3Aopen+is%3Aissue) that reports the same bug
- [X] I've seen the [troubleshooting guide](https://docs.continue.dev/troubleshooting) on the Continue Docs

### Relevant environment info

```Markdown
- OS:MacOS
- Continue version: 0.9.244
- IDE version: VS Code 1.96.0
- Model:
- config.json:
  
    {
      "title": "Open Router",
      "provider": "openrouter",
      // Non Working models
      // "model": "meta-llama/llama-3.3-70b-instruct",
      // "model": "cohere/command-r7b-12-2024",
      // "model": "qwen/qwen-2.5-coder-32b-instruct",
      // "model": "google/gemini-exp-1121:free",
      // "model": "nvidia/llama-3.1-nemotron-70b-instruct",

      // Working models
      // "model": "anthropic/claude-3.5-sonnet:beta",
      // "model": "openai/gpt-4o-2024-11-20",
      // "model": "mistralai/mistral-large-2411",
      // "model": "x-ai/grok-beta",
      // "model": "amazon/nova-lite-v1",

      "apiBase": "https://openrouter.ai/api/v1",
      "apiKey": "API Key"
    },
```

### Description

Chat doesn't work with the following test providers/models due unsupported tools context option:

- "meta-llama/llama-3.3-70b-instruct"
- "cohere/command-r7b-12-2024",
- "qwen/qwen-2.5-coder-32b-instruct",
- "google/gemini-exp-1121:free",
- "nvidia/llama-3.1-nemotron-70b-instruct",

### To reproduce

Switch the model and start chat

### Log output

```Shell
console.ts:137 [Extension Host] Error handling webview message: {
  "msg": {
    "messageId": "fde3d646-dcf7-4652-aafa-7e254f8f85b8",
    "messageType": "llm/streamChat",
    "data": {
      "messages": [
        {
          "role": "user",
          "content": [
            {
              "type": "text",
              "text": "hello"
            }
          ]
        },
        {
          "role": "assistant",
          "content": ""
        }
      ],
      "title": "Open Router",
      "completionOptions": {
        "tools": [
          {
            "type": "function",
            "displayTitle": "Read File",
            "wouldLikeTo": "read {{{ filepath }}}",
            "readonly": true,
            "function": {
              "name": "builtin_read_file",
              "description": "Use this tool whenever you need to view the contents of a file.",
              "parameters": {
                "type": "object",
                "required": [
                  "filepath"
                ],
                "properties": {
                  "filepath": {
                    "type": "string",
                    "description": "The path of the file to read, relative to the root of the workspace."
                  }
                }
              }
            }
          },
          {
            "type": "function",
            "displayTitle": "Create New File",
            "wouldLikeTo": "create a new file",
            "readonly": false,
            "function": {
              "name": "builtin_create_new_file",
              "description": "Create a new file",
              "parameters": {
                "type": "object",
                "required": [
                  "filepath",
                  "contents"
                ],
                "properties": {
                  "filepath": {
                    "type": "string",
                    "description": "The path where the new file should be created"
                  },
                  "contents": {
                    "type": "string",
                    "description": "The contents to write to the new file"
                  }
                }
              }
            }
          },
          {
            "type": "function",
            "displayTitle": "Run Terminal Command",
            "wouldLikeTo": "run a terminal command",
            "readonly": false,
            "function": {
              "name": "builtin_run_terminal_command",
              "description": "Run a terminal command in the current directory. The shell is not stateful and will not remember any previous commands.",
              "parameters": {
                "type": "object",
                "required": [
                  "command"
                ],
                "properties": {
                  "command": {
                    "type": "string",
                    "description": "The command to run. This will be passed directly into the shell."
                  }
                }
              }
            }
          },
          {
            "type": "function",
            "displayTitle": "View Subdirectory",
            "wouldLikeTo": "view the contents of \"{{{ directory_path }}}\"",
            "readonly": true,
            "function": {
              "name": "builtin_view_subdirectory",
              "description": "View the contents of a subdirectory",
              "parameters": {
                "type": "object",
                "required": [
                  "directory_path"
                ],
                "properties": {
                  "directory_path": {
                    "type": "string",
                    "description": "The path of the subdirectory to view, relative to the root of the workspace"
                  }
                }
              }
            }
          },
          {
            "type": "function",
            "displayTitle": "View Repo Map",
            "wouldLikeTo": "view the repository map",
            "readonly": true,
            "function": {
              "name": "builtin_view_repo_map",
              "description": "View the repository map",
              "parameters": {
                "type": "object",
                "properties": {}
              }
            }
          },
          {
            "type": "function",
            "displayTitle": "Exact Search",
            "wouldLikeTo": "search for \"{{{ query }}}\" in the repository",
            "readonly": true,
            "function": {
              "name": "builtin_exact_search",
              "description": "Perform an exact search over the repository using ripgrep.",
              "parameters": {
                "type": "object",
                "required": [
                  "query"
                ],
                "properties": {
                  "query": {
                    "type": "string",
                    "description": "The search query to use. Must be a valid ripgrep regex expression, escaped where needed"
                  }
                }
              }
            }
          },
          {
            "type": "function",
            "displayTitle": "Search Web",
            "wouldLikeTo": "search the web for \"{{{ query }}}\"",
            "readonly": true,
            "function": {
              "name": "builtin_search_web",
              "description": "Performs a web search, returning top results. This tool should only be called for questions that require external knowledge. Common programming questions do not require web search.",
              "parameters": {
                "type": "object",
                "required": [
                  "query"
                ],
                "properties": {
                  "repo_url": {
                    "type": "string",
                    "description": "The natural language search query"
                  }
                }
              }
            }
          },
          {
            "type": "function",
            "displayTitle": "View Diff",
            "wouldLikeTo": "view a diff",
            "readonly": true,
            "function": {
              "name": "builtin_view_diff",
              "description": "View the current diff of working changes",
              "parameters": {
                "type": "object",
                "properties": {}
              }
            }
          },
          {
            "type": "function",
            "displayTitle": "Read Currently Open File",
            "wouldLikeTo": "read the current file",
            "readonly": true,
            "function": {
              "name": "builtin_read_currently_open_file",
              "description": "Read the currently open file in the IDE. If the user seems to be referring to a file that you can't see, this is probably it.",
              "parameters": {
                "type": "object",
                "properties": {}
              }
            }
          }
        ]
      }
    }
  }
}

Error: HTTP 404 Not Found from https://openrouter.ai/api/v1/chat/completions

{"error":{"message":"No endpoints found that support tool use. To learn more about provider routing, visit: https://openrouter.ai/docs/provider-routing","code":404}} (at console.<anonymous> (file:///Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:159:26388))
```

**Comment by C-W-D-Harshit** [comment]
+1

**Comment by Ricardomol** [comment]
+1

**Comment by phpmac** [comment]
fuck

**Comment by blissiraDC** [comment]
+1

**Comment by iorveth** [comment]
+1

**Comment by RageshAntonyHM** [comment]
Also occurs in deepseek/deepseek-r1:nitro

**Comment by iFarmGolems** [comment]
+1

**Comment by FelipeVeiga** [comment]
+1

**Comment by gkonuralp** [comment]
Just reinstalled the extension and this is the situation now. Prior to that it was working.

**Comment by ShellOliver** [comment]
same

**Comment by raphael-foliveira** [comment]
+1

**Comment by tomkat-cr** [comment]
same here... any thoughts on this issue ?

**Comment by mntdherm** [comment]
+1

**Comment by sestinj** [comment]
Hey all! Thank you for all the +1s. This issue should now be resolved in the latest pre-release for VS Code. Would appreciate verification of this—I won't close the issue until I am sure everything is back to normal, but in my own testing with these models it seems to work

**Comment by mntdherm** [comment]
> Hey all! Thank you for all the +1s. This issue should now be resolved in the latest pre-release for VS Code. Would appreciate verification of this—I won't close the issue until I am sure everything is back to normal, but in my own testing with these models it seems to work

still some of them not working, like deepseek coder

**Comment by Sanerok** [comment]
`Error code: 404 - {'error': {'message': 'No endpoints found that support tool use. To learn more about provider routing, visit: https://openrouter.ai/docs/provider-routing', 'code': 404}` with `qwen/qwen-2-vl-7b-instruct`

**Comment by 0x0642ndt** [comment]
+1 with `deepseek/deepseek-r1`, although used `get_supported_openai_params` with this link [Input Params
](https://docs.litellm.ai/docs/completion/input)

**Comment by jamesbanksco** [comment]
+1

**Comment by nabang1010** [comment]
+1

**Comment by AdamerTooploox** [comment]
+1

**Comment by kunal5** [comment]
+1 same here with `deepseek/deepseek-r1-zero:free`

**Comment by vitovtnet** [comment]
+1 same with a lot of model, even with mistralai/mistral-nemo:free

**Comment by lifanwar** [comment]
That's same with all of perplexity model

**Comment by jhonsfran1165** [comment]
+1 x-ai/grok-2-vision-1212

**Comment by dodandeniya** [comment]
+1 qwen/qwen3-30b-a3b:free

**Comment by sparkncraft** [comment]
+ 1

**Comment by lydiaBn** [comment]
+1

**Comment by koshibatRL** [comment]
+1 same here with `qwen/qwen3-4b:free`

**Comment by porcorreo** [comment]
+1

**Comment by truongnmt** [comment]
+1

**Comment by truongnmt** [comment]
I think it is a problem with openrouter, not sure what we can do here

**Comment by simon28082** [comment]
+1

**Comment by UsamaEjaz0** [comment]
+1

**Comment by rontrandev** [comment]
+1

**Comment by sestinj** [comment]
Hey all—want to clarify a few things and share our upcoming solution.

These models should all work in "Chat" mode (if not, please let me know and this might be a separate issue), but for models that aren't working in "Agent" mode, this is because OpenRouter does not support tool use. Other tools that allow tool use through OpenRouter aren't actually using native tools, they are putting all of the tools into the prompt.

We are working on accomplishing this in an upcoming PR for ["Non-native tool support"](https://github.com/continuedev/continue/pull/5809). Once this is merged, OpenRouter should work with any model!

**Comment by The-Best-Codes** [comment]
This is an OpenRouter issue. OpenRouter routes your request to any provider it can find (which is why it has such good uptime), but if none of the providers it finds for a model are marked as supporting tool use, it throws this error.

**Comment by BradKML** [comment]
@The-Best-Codes, the same would apply for niche providers. Can't some parameters be specifically set for filtering providers supporting tool use?

**Comment by The-Best-Codes** [comment]
@BradKML You can search for models supporting tool use on the OpenRouter website, although some of the ones I try here don't actually have a provider (they are just deprecated).

https://openrouter.ai/models?fmt=cards&supported_parameters=tools

**Comment by pierrecorsini** [comment]
> [@BradKML](https://github.com/BradKML) You can search for models supporting tool use on the OpenRouter website, although some of the ones I try here don't actually have a provider (they are just deprecated).
> 
> https://openrouter.ai/models?fmt=cards&supported_parameters=tools

But has it worked for you ? I did the test with `deepseek/deepseek-chat-v3-0324` that is classified as "supported parameters: tools", works, and has plenty of providers; the problem persists.

**Comment by The-Best-Codes** [comment]
@pierrecorsini Yes, it works for me as far as tool use. The only issue I've found is that with older models, there are no longer any providers period. So the model just doesn't work at all.

**Comment by ser888gio** [comment]
Tried with ElevenLabs and unfortunately it did not work.

**Comment by aknahin** [comment]
Using in n8n, model I was trying to use is `perplexity/sonar` and it shows this error.

**Comment by EgorLarkin** [comment]
+1 
https://openrouter.ai/api/v1 - says, that model is not available
